docs/nebula-design-concepts.md

```diff
@@ -0,0 +1,211 @@
+multiple storage engines like Rocksdb and HBase, etc.
+
+Raft achieves data consensus via an elected leader. Based on that, nebula-storaged makes the following optimizations:
+
+- Parallel Raft
+
+      Partitions of the same ID from multiple machines form a raft group. And the parallel operations are implemented with multiple sets of Raft groups.
+
+- Write Path & batch
+
+      In Raft protocol, the master replicates log entries to all the followers and commits the entries in order. To improve throughput, Nebula Graph makes
+      one host hold many raft groups, and the operations on different group could be invoked simultaneously.
+
+- Add Raft Learner
+
+      When a new server joins the cluster, it can be added as a learner, which can neither vote nor count towards quorum. Once it catches up to the leader's logs, it can be promoted to follower as a normal voting node.
+
+- Load-balance
+
+      Migrating the partitions on an overworked server to other relatively idle servers to increases availability and capacity of the system.
+
+
+
+![image](https://user-images.githubusercontent.com/42762957/61120371-6f0fc200-a4cf-11e9-8c41-9e531380205b.png)
+
+
+
+### Metadata service
+
+The process corresponding to the meta service is nebula-metad, and its main functions are:
+
+-  User management
+
+      In Nebula Graph different roles are assigned diverse privileges. We provide the following native roles: God user, Admin, User and Guest.
+
+- Cluster configuration management
+
+      Metadata service manages the servers and partitions in the cluster, e.g. records location of the partitions, receives heartbeat from servers, etc. It balances the partitions and manages the communication traffic in case of server failure.
+
+- Graph space management
+
+      Metadata service stores the metadata of all spaces in the cluster and tracks changes that take place in these spaces, like adding, dropping space, modifying graph space configuration (Raft copies).
+
+-   Schema management
+
+      Nebula Graph is a strong typed database.
+
+      - Types of tag and edge properties are recorded by meta service. Supported data types are: int, double, timestamp, list, etc.
+
+      - Multi-version management, supporting adding, modifying and deleting schema, and recording its version.
+
+      - TTL (time-to-live) management, supporting automatic data deletion and space reclamation.
+
+The meta service is stateful, and just like the storage service, it persists data to a key-value store.
+
+
+![image](https://user-images.githubusercontent.com/42762957/61120413-8cdd2700-a4cf-11e9-8846-14b5d8bd6693.png)
+
+
+### Query Engine & Query Language (nGQL)
+
+The process corresponding to the query engine is nebula-graphd, which consists
```

Comment:
The binary of the query engine is nebula-graphd. Each nebula-graphd instance is stateless and never talks to other nebula-graphd. nebula-graphd only talks to the storage service and the meta service. That makes it trivial to expand or shrink the query engine cluster.